scons: Change to Kconfig build system

rogerchang23424 opened the pull request:
The PR contains the following changes:
- Move all of the config options(`env["CONF"]`) from SConsopt to Kconfig files
- Update `build_opts` files to Kconfig option formats
- The Ruby Protocol files are only built if `RUBY=y`
- Remove the default-default build target
- Kconfig commands are included in the PR:
    - defconfig
    - setconfig
    - meunconfig
    - guiconfig
    - listnewconfig
    - savedefconfig
    - oldconfig
    - olddefconfig
- Add the `python3-tk` package dependencies
 
Jira issue: https://gem5.atlassian.net/browse/GEM5-1211

Commit:
scons: Use either the "build" or "gem5.build" as build anchor.

If gem5.build already exists within a directory, then that build
directory can be used without having to worry about variants.

If it doesn't exist and we find a build/${VARIANT} style path, then we
use that as the anchor.

In either case, the variant name is the final component of the build
path. The parse_build_path function had been separating that out, but it
was just put back onto the path again anyway by the only caller, and
then split out again when that path was consumed. We save a step by not
splitting it out in parse_build_path.

Change-Id: I8705b3dbb7664748f5525869cb188df70319d403

Commit:
scons: Use Kconfig to configure gem5.

These are not yet consumed by anything, but convert all the settings
from SCons variables to Kconfig variables.

If you have existing SConsopts files which need to be converted, you
should take a look at KCONFIG.md to learn about how kconfig is used in
gem5. You should decide if any variables need to be available to C++ or
kconfig itself, and whether those are options which should be detected
automatically, or should be up to the user. Options which should be
measured automatically should still be in SConsopts files, while user
facing options should be added to new or existing Kconfig files.

Generally, make sure you're storing c++/kconfig visible options in
env['CONF'][...]. Also remove references to sticky_vars since persistent
options should now be handled with kconfig, and export_vars since
everything in env['CONF'] is now exported automatically.

Switch SCons/gem5 to use Kconfig for configuration, except EXTRAS which
is still a sticky SCons variable. This is necessary because EXTRAS also
controls what config options exist. If it came from Kconfig itself, then
there would be a circular dependency. This dependency could
theoretically be handled by reparsing the Kconfig when EXTRAS
directories were added or removed, but that would be complicated, and
isn't supported by kconfiglib. It wouldn't be worth the significant
effort it would take to add it, just to use Kconfig more purely.

Change-Id: I29ab1940b2d7b0e6635a490452d05befe5b4a2c9

Commit:
scons: Add a build script for generating a root Kconfig file.

This root Kconfig file "source"s (includes) the base gem5 src/Kconfig
file, and also any optional Kconfig files found in the base of EXTRAS
directories. These will be called out in the menuconfig interface and
config files with the name of the EXTRAS directory they came from, and a
blank section will be present either if the Kconfig didn't exist, or it
did exist but had no options in it.

Change-Id: I54060d613f0e0ab9372bed37a2fe5849bf5bbcdb

Commit:
scons: Add a mechanism to run menuconfig to set up a build dir.

If you call scons with the fist argument set to menuconfig, that means
to run menuconfig on the path following it. Or in other words, if
you ran this command:

scons menuconfig build/foo/bar

That would tell SCons to set up a build directory at the path
build/foo/bar, and then invoke menuconfig so you can set up its
configuration.

In addition to using this mechanism to set up a new build directory, you
can also use it to reconfigure an existing directory.

This supplements and does not replace the existing mechanism of using
"build/${VARIANT}" to select a config with defconfig.

Change-Id: Ief8e8c2ee6477799455c2004bef06c64be5cc1db

Commit:
scons: Add a mechanism to manually defconfig a build dir.

This will let you specify *any* defconfig file, instead of implicitly
selecting one from the defconfig directory based on the variant name.

Change-Id: I74c981b206849f08e60c2df702c06534c670cc7c

Commit:
scons: Hook in the kconfig setconfig utility.

This little utility lets you set particular values in an existing config
without having to open up the whole menuconfig interface.

Also reorganize things in kconfig.py a little to help share code between
wrappers.

Change-Id: I7cba0c0ef8d318d6c39e49c779ebb2bbdc3d94c8

Commit:
scons: Hook in the listnewconfig kconfig helper.

This helper lists config options which are new in the Kconfig and which
are not currently set in the config file.

Change-Id: I0c426d85c0cf0d2bdbac599845669165285a82a0

Commit:
scons: Hook up the kconfig guiconfig program.

Change-Id: I0563a2fb2d79cea5974aeaf65a400be5ee51dc63

Commit:
scons: Hook up the savedefconfig kconfig helper.

This helper utility lets you save the defconfig which would give rise to
a given config. For instance, you could use menuconfig to set up a
config how you want it with the options you cared about configured, and
then use savedefconfig to save a defconfig of that somewhere to the
side, in the gem5 defconfig directory, etc. Then later, you could use
that defconfig to set up a new build directory with that same config,
even if the kconfig options have changed a little bit since then.

A saved defconfig like that can also be a good way to visually see what
options have been set to something interesting, and an easier way to
pass a config to someone else to use, to put in bug reports, etc.

Change-Id: Ifd344278638c59b48c261b36058832034c009c78

Commit:
scons: Hook up oldconfig and olddefconfig.

These two utilities help update an old config to add settings for new
config options. The difference between them is that oldconfig asks what
new settings you want to use, while olddefconfig automatically picks the
defaults.

Change-Id: Icd3e57f834684e620705beb884faa5b6e2cc7baa

Commit:
scons: Flesh out the help text for "magic" targets.

These targets are not necessarily obvious, and tell SCons to do useful
things, like build a particular version of the gem5 binary with a
particular configuration, or run the unit tests.

Add descriptions of these targets to the help so that they are much
more discoverable.

Change-Id: If84399be1a7155ff5f66f511efe1f1c241089c84

Commit:
scons: Remove the default-default build target.

In gem5, there are many equally valid and equally useful top level
targets which the user might want. It no longer makes sense to
arbitrarily pick one to be the default target. It makes sense to force
the user to actually specify what they want, instead of assuming it
must be the ARM debug binary.

There is currently an M5_DEFAULT_BINARY environment variable which
will change what the default binary is, if set. This change leaves
that in place, but removes the default-default, or in other words the
default that is used if M5_DEFAULT_BINARY is not set.

This way if the user knows what default they want, they can specify it
locally in their environment and avoid having to type it over and over
again, but we're not making an arbitrary choice at a more global level
without the context to know what actually makes sense.

Change-Id: I886adb1289b9879d53387250f950909a4809ed8b

rogerchang23424 commented:
Previous patch discussion: https://gem5-review.googlesource.com/c/public/gem5/+/56590
Jira issue: https://gem5.atlassian.net/browse/GEM5-1211

powerjg commented:
I pinged all maintainers on this as it's a major change to the way gem5 is configured at build time and it's a major change to how users build gem5. We should all take a close look at this "API" breaking change to make sure that it's not going to affect the gem5 users' workflows.

Below is an excerpt of a conversation that Roger and I had over email for other developers to see. (Bracketed text is mine.) This includes some answers to how this is changing gem5's build API.

----- 

## For users:

### How do I build gem5?

All of the available variants are listed in the “defconfig” directory instead of in “build_opts”.  You need to use the new build scheme to build gem5. Please follow the steps below.

Example:

Initialize gem5 build

**[NOTE: THIS IS A MAJOR CHANGE]**

```
$ scons defconfig gem5_build defconfig/<variant>
```

The command creates a new gem5.build directory in the build directory you specified, the gem5.build includes the Kconfig setting and global config headers.

Build gem5 instance

**[NOTE: THIS IS A MAJOR CHANGE]**

```
$ scons gem5_build/gem5.opt
```

The command builds the gem5 instance based on the Kconfig in the gem5.build.

We also support old build schemes for backward compatibility. The following commands are equivalent.

Old build schemes

```
$ scons build/<variant>/gem5.opt
```

New build schemes
```
$ scons defconfig build/<variant> defconfig/<variant>
$ scons build/<variant>/gem5.opt
```

### How do I customize the gem5 build?

**[NOTE: THIS IS A MAJOR CHANGE]**

Users can use `scons setconfig` to change the config from the command line or use `scons menuconfig` to change the config with UI. Here's an example.

If users want to enable kvm support based on ARM defconfig, we can do the following commands.

setconfig:
```
$ scons setconfig gem5_build USE_KVM=y
$ scons gem5_build/gem5.opt
```

menuconfig (will list all of available build options):
```
 $ scons menuconfig
 ```

### How do I find out what gem5's build-time options are?

**[NOTE: THIS IS A MAJOR CHANGE]**

Users can run `scons menuconfig` to view available options. If the gem5 binary is built, they will show the variables in `buildEnv` in `${BUILDDIR}/python/m5/defines.py`

## For developers:

### How do I extend the gem5 build with new build-time options? 

After configuring the options with KConfig, the build options would be available in `env['CONF']` in Scons scripts. If the user want to build some files only if options enabled, they can add following code in the SConscript:

```
If env[‘CONF’][<option>]:
env.TagImplies(<tag for option>, ‘gem5 lib’)
```

Where <option> is the option name in the KConfig files, <tag for option> is the tag option related to the <option> from KConfig files which is used to determine the file be compiled, <tag for option> name can be different with <option>.

If the source file want to build only if specified option enabled, write like:

```
Source(‘<any filename>[.cc](http://foo.cc/)’, tags=<tag for option>)
```

Where <any filename> can be anything you want.

### How do I modify gem5's build-time options (e.g., remove an option)?

There are two way to define the build-time options

Configurable options: defined by KConfig

Global options: defined by Sconsopt

Developers should reach the correct files depending on the options they want to modify.

After modification and reconfiguring by `scons defconfig`, the build options will be in the env[‘CONF’] for scons scripts as mentioned before..   Here's an example to add an option in KConfig.

```
FOO
<type> <description of FOO>
Default xxx
```

For more information about how to add options to KConfig, please refer to the document: https://www.kernel.org/doc/html/latest/kbuild/kconfig-language.html

powerjg commented:
@rogerchang23424 

A few more follow up questions.

First, it would also be helpful to explain what the benefits of this change are. What is the problem that this change is solving? Or, why is this better than the previous implementation which was a simple flat file with configuration options that could be overridden on the command line when building? Or, what new features does this change enable that could not be done before?
I appreciate the contribution, but given that it's not a trivial change for developers or users, I want to make sure that it's a positive change for the gem5 community.

Below are some specific follow ups to the answers you gave above:

1. For the "backwards compatible" build scheme...
  a. What is the *right* way to build gem5 now? I.e., when a new user asks "how do I build gem5?" what do we say? Do we say `scons build/<variant>/gem5.opt`, or do we say "use menuconfig", or do we say "use defconfig." It is very important for the community to understand what the supported build commands will be going forward.
  b. Where do I find "\<variant\>"? Is this now in the `defconfig` directory? Why did this directory change names from `build_opts`?
2. Under "how do I extend gem5 build with new build-time options" you said "After configuring the options with KConfig..."
  a. How do I configure an option with KConfig? What file do I modify? How do I modify that file?
  b. What is "\<tag for option\>" for the `Source` command in SCons? Is this the same as the `<option>` or something different? How do I specify that tag?
3. Under "how do I modify gem5's build-time options..." you said "Developers should reach the correct files". How do I know what the correct files are? How do I know what options are in KConfig vs global options defined by SConsopt?

rogerchang23424 commented:
Previous patch discussion: https://gem5-review.googlesource.com/c/public/gem5/+/56590

CC: @wmin0, @EarlOu

andysan commented:
I think there are a couple of obvious benefits to users, especially in complex environments:

- It makes it easier to discover the available build-time configuration options. Each option would be visible in the configuration GUI and the Kconfig language encourages better documentation. None of this is currently possible / easy.
- More obvious dependency tracking. The current flat doesn't make dependencies obvious. In most cases, options just silently don't work if they aren't supported for a specific configuration.
- It makes it easier to build custom configurations in multi-ISA systems where it is desirable to compile a subset of the available ISAs.
- It enables complex build-time configuration that is required for hybrid environment where gem5 is bolted to an existing simulator.
- It would make it easier to reason about ISAs and device dependencies. We currently put some generic devices in `src/dev/{ARCH}/`. Some of this could be cleaner in Kconfig since ISAs could set configuration variables that are dependencies for specific devices. For example, a platform could define `HAVE_BUS_PCI` and the corresponding device would depend on this high-level property instead of "I know that I'm on an Arm platform which means that these devices should be enabled".

Being able to strip down the build configuration would be quite useful in some cases as well. For example, if I only want to use SE mode, why would I build the entire PCI subsystem? This use case would require some additional work, but having support for Kconfig would be a very good start.

Overall, I think this is a positive change that enables a lot of useful functionality. We just need to make sure that new users have a simple way of building gem5 with sensible default configurations which was possible last time I reviewed this.

powerjg commented:
I think everything you say makes sense, Andreas. One thing I'll point out is that all of the benefits you list require that we do a lot of new education for our users. For instance, if this was committed as is, there would be no way for users to discover that `scons menuconfig` existed if they continue to use `scons build/<variant>/gem5.opt`. To achieve the benefits that you enumerate is going to require a lot of documentation updates and a lot of changes to the tutorials that we run. 

As I like to say when I review papers, "the impact of this change would be much larger if" it also included updates to educate gem5's users. Without including these education updates (e.g., a blog post, documentation on the website, updates to tutorial information, etc.), the positive benefits that Andreas listed are unlikely to be realized by most gem5 users.

rogerchang23424 commented:
Hi Jason and all,

Apologize to reply the comment lately.

Let me answer the two topics first, and then the specific questions later:

1. What's the benefit of this change?

    a. The change cleans up the existing configuration method. In the original flow, some explicit configuration variables are sticky and some are not, some will be exported to c++ code and some will not. This causes ambiguity and sometimes the same command will generate different results depending on if it's the first time using (due to the sticky) or not.

         1. In the new flow, all explicit configuration variables will become sticky, or will be demoted to environment variables. As part of making sticky variables clear, all explicit configuration now requires 2 steps or editing a config file instead of using the command line. This process makes it clear that those are "sticky".
         2. All configuration variables will be exported. It will be up to gem5 source files to decide what variables they actually want, and if there is no dependency on them
         3. After the change, it will be clear that we have 3 different types of configs:
                    i. Configuration option - persist (sticky), need to be written with editor, and affect build product
                    ii. Env variable - not persist, set with command line, and affect build product
                    iii. Command line option - not persist, set with command line, and affect build process

   b. As part of the benefit from a), now a user can use `scons menuconfig <build_folder>` to see and edit all available options. In the past, there was no way to list all possible options.

   c. The new flow allows a user to fine-grand define what to be included in the final result. For example, if we'd like to work on a SoC platform with two types of ISA CPU cores, the existing flow only has the "ALL" option. The new flow allows us to only pick the two ISAs we want. In our use case, due to the large number of IP models we have, we've found this to be useful to minimize the build time and build size.

   d. Some options should only be set when the specified options are enabled. For example, `PROTOCOL` should be specified only if `RUBY=y`, otherwise `PROTOCOL` is always empty whatever the developers try to set `PROTOCOL` in the variant files. For another example, `USE_ARM_FASTMODEL` can be enabled if `USE_ARM_ISA` and `USE_SYSTEMC` are enabled, otherwise, `USE_ARM_FASTMODEL` is always disabled.

2. What's the suggested way for new users to build gem5?

    As suggested from the previous review, we'd like to make it easy for most of the users. After some internal discussion, we'd like to see if the following documentation proposal is OK for the reviewers:
    - For typical users: use `scons build/<variant>/gem5.opt` to build
    - For advance users who want to change their default build options:
           1. `scons defconfig <build_folder> build_opts/<variant>`
           2. `scons menuconfig <build_folder>`
           3. `scons <build_folder>/gem5.opt`


For the specific questions from Jason:

1.  For the "backwards compatible" build scheme...
                 a. Please see above answers on the suggested way to build gem5.
                 b. We decide to move all the variants back to `build_opts`. So old developers can still find them.

2. Under "how do I extend gem5 build with new build-time options" you said "After configuring the options with KConfig..."
    a. To add a new option, one can add the option in the Kconfig closer to the source, e.g. USE_X86_ISA is in `src/arch/x86/Kconfig`. And then use the option in the corresponding Sconscript, e.g. `src/arch/x86/Sconscript` as mentioned in the previous answer. 
    b. The [`Source`](https://github.com/gem5/gem5/blob/stable/site_scons/gem5_scons/sources.py#L246) action in Gem5 Scons has tags attribute which make developers can select a group of actions. By default Gem5 build with [`gem5 lib`](https://github.com/gem5/gem5/blob/stable/src/SConscript#L717). In addition, developers can use [`TagImplies`](https://github.com/gem5/gem5/blob/stable/site_scons/gem5_scons/sources.py#L51) to associate two tag. So we can use the tags to implement optional build.

```python
if env['CONF']['FOO']:
   env.TagImplies('foo', 'gem5 lib')
Source('foo.cc', tags='foo')
```

However we think using tag to implement the optional build is too complicated for normal users. We would like to suggest in another simpler way. We can just check if the config is enabled or just return in the `SConscript`.

```python
if not env['CONF']['FOO']::
   Return()
```

3.  Under "how do I modify gem5's build-time options..."
Most of the options are still in the scons files. For the configuration options originating from `sticky_vars`, such as `USE_ARM_ISA`, `USE_RISCV_ISA`, `PROTOCOL`, `BUILD_GPU` and so on. They will be moved to the Kconfig file from Sconsopts in the same directory. Developers can find such options in the Kconfig file in the same directory. The users can also use `menuconfig` command to know the detail about configuration options

Type `scons menuconfig <builddir> to enter the menu config UI
 ![image1](https://github.com/gem5/gem5/assets/32214817/c041a243-b3e2-4330-a48c-831b5650ef8c)

To show the configuration option name, press ‘c’ button
![image3](https://github.com/gem5/gem5/assets/32214817/57717135-3495-46ed-aaa6-67ba2d00c382)

To know the details about options, press ‘?’ to the specified option. Here I take the `USE_ARM_ISA` as an example.
![image2](https://github.com/gem5/gem5/assets/32214817/9700712e-9de1-449f-826e-bef58cd17480)
It includes where the configuration options come from. What the kconfig gets the options and so on.

For the guiconfig, it will open the new windows to list all of the configuration options
![image4](https://github.com/gem5/gem5/assets/32214817/6d4b352a-3d57-49cf-be89-2c81a348c502)

rogerchang23424 commented:
@powerjg  

> I think everything you say makes sense, Andreas. One thing I'll point out is that all of the benefits you list require that we do a lot of new education for our users. For instance, if this was committed as is, there would be no way for users to discover that `scons menuconfig` existed if they continue to use `scons build/<variant>/gem5.opt`. To achieve the benefits that you enumerate is going to require a lot of documentation updates and a lot of changes to the tutorials that we run.
> 
> As I like to say when I review papers, "the impact of this change would be much larger if" it also included updates to educate gem5's users. Without including these education updates (e.g., a blog post, documentation on the website, updates to tutorial information, etc.), the positive benefits that Andreas listed are unlikely to be realized by most gem5 users.

We will let typical user continue to use `scons build/<variant>/gem5.opt` to build gem5. I think it's not required to update the tutorial document. In the other hand, we can provide the guideline about Kconfig for advanced user.

EarlOu commented:
I can see a lot of good questions from Jason about documentation, and also a lot of them should be answered by documentation instead of review comment (so the future users can get same information on a nice page).

So I'd like to propose we review in the following way:

- Roger should upload another RP on the website repo (https://github.com/gem5/website) to document the Kconfig usage. And we review the documentation there.
- In this PR, we review the design and implementation to see if everyone agree this is a good change to gem5.

In the document review, most of the question should be answered by new patch in the PR instead of review comment to make sure future users get the right information.

powerjg commented:
Thanks for all of the great information, Roger! Adding this documentation to the website will be a great way to alleviate many of my concerns. Apologies for the wall of text below. I am not really looking for any specific responses to questions, but I want to give you my perspective on this change and why I've been pushing back.

I think that your 1a benefit (that this change makes all variables "sticky") is something that greatly improves the maintainability of the gem5 build system and is a huge improvement. I am strongly in favor of this change! I'm a bit disappointed that it comes with the added complexity of KConfig, though. From my point of view, these two changes, reducing the complexity of the build-time options and KConfig, should be orthogonal. The only part of this change that I'm pushing back on is adding support for KConfig.

I'm not going to unilaterally block this change, but I want to point out two worries that I have with having to maintain this going forward:
(1) When helping people that have problems with gem5, it is going to become increasingly difficult to reproduce their errors if they use a highly customized gem5 build. Additionally, we may find that there are ways to configure gem5 with KConfig that (a) are never tested and (b) are broken/unsupported. This isn't a problem for expert users like you all, but we have many users that are not as expert as you at building and maintaining Linux-based applications. I am strongly against saying "we won't answer questions on gem5 if you use a customized build." Instead, as maintainers, I would like to see us make it *easier* to provide help to users.
(2) If we ever want to move away from scons (which I think we do!), we are now going to have to find a build system that supports KConfig. I see that zephyr seems to supports cmake+KConfig through their own customization, but their build flow is incredibly complicated! (See this link https://docs.zephyrproject.org/latest/_images/build-config-phase.svg). I worry that building our own new customization to use KConfig on top of some other build system is going to be a large and costly undertaking which will make it more difficult to ever move away from scons.

Having worked with gem5 for the past 12 years and been one of the lead maintainers for the past 6 years, I have found the most difficult parts of gem5 to maintain is when we have non-standard implementations and non-standard use cases for things. SCons is one of the most difficult parts of the project to maintain, and I am afraid that this added customization is going to make it even more difficult to maintain going forward. Furthermore, the way that we have customized things makes it very difficult to find answers on Google when something goes wrong, which further increases the burden on the community to help users and makes maintenance more difficult.

---

As a side note and a potential data point that may help you understand where I'm coming from... I am very familiar with how to use KConfig. I've been compiling my own kernels for the past 20 years. So for me, I would have no problem using this new system, and I totally see it's benefits for people who are highly familiar with how to build and deploy Linux applications.
However, 90% of the students that I work with in my group, 90% of the students that attend the gem5 tutorials, 90% of the students that attended the gem5 bootcamp last year... they had never compiled a kernel and had never used KConfig. In fact, for many gem5 users, building gem5 is one of the first times they have ever built an application on the command line! By adding this as another dependency to use gem5, we are raising the barrier to entry and essentially gatekeeping people coming into computer architecture research. My personal goal (as a professor and as one of the leaders of this community) is to make computer architecture research *easier* and to bring more people into the community. 
I hear you when you say "well, you can just build gem5 the old way, still." However, this isn't the whole problem, as I've pointed out above. The community is going to have to support users in using KConfig and support cases where things go wrong. This change is also going to make it more difficult to move to a better-known build system than our highly-customized scons implementation.
In general, I'm going to push back when I see changes that are increasing the difficulty in getting started with gem5. I know this isn't everyone's goals with this project, but I wanted to give you some of my perspective.

wmin0 commented:
Hi Jason,

Thanks for explaining the concern. I would like to try to answer them.

>  it is going to become increasingly difficult to reproduce their errors if they use a highly customized gem5 build. Additionally

Actually current SCons has supported customized build already. For example you can enable the SystemC integration by the command below.

```
$ python3 /usr/bin/scons -C gem5 USE_SYSTEMC=1 build/ARM/gem5.opt
```

If you dig into the build directory `build/ARM/gem5.build`, there's no record about how you built. `build/ARM/gem5.build/variable` only contains the part in build_opts. It's the current difficulty to reproduce the build.

With KConfig we put all variables in `build/ARM/gem5.build/config`. People can just provide the config file to us to reproduce. I would say it's a cleanup of the way defining build variables. I think it would make maintainers providing help to users easier.

> I worry that building our own new customization to use KConfig on top of some other build system is going to be a large and costly undertaking which will make it more difficult to ever move away from scons.

The issue we met today is because SCons doesn't native support such complicated build scheme. Therefore, we did lots of works in [SConstruct](https://github.com/gem5/gem5/blob/develop/SConstruct) and [site_scons](https://github.com/gem5/gem5/tree/develop/site_scons). To move away from Scons, we need to extract the works we did from current SCons. Previously we have already moved the codegen part into individual python tools. This time we want to extract the build variables from SCons. Moving configuration step out of SCons doesn't increase the difficulty of moving away from SCons. Instead, it provides the opportunity to share the config between different build system. We can develop a parallel build system with the same config and then get away from SCons.

I think zephyr is not a very good example because it contains too many different things to build in the build flow. In gem5 case, device tree is apart from the main build process [dt](https://github.com/gem5/gem5/tree/develop/system/arm/dt).

Finally, I would like to mention again that the customized build is already enabling in current SCons but seems no document talking about that. The KConfig feature is just bring that into users' eyes and making it more easily to use. We can also hide the capability but I don't think you want going that way.

I'm not sure if I answer parts of your concern. Please tell me any thought :)

Thanks.

powerjg commented:
> If you dig into the build directory `build/ARM/gem5.build`, there's no record about how you built. `build/ARM/gem5.build/variable` only contains the part in build_opts. It's the current difficulty to reproduce the build.

I strongly agree that this is a problem that should be solved! Fixing this is a big improvement to gem5.

> The issue we met today is because SCons doesn't native support such complicated build scheme. Therefore, we did lots of works in [SConstruct](https://github.com/gem5/gem5/blob/develop/SConstruct) and [site_scons](https://github.com/gem5/gem5/tree/develop/site_scons). To move away from Scons, we need to extract the works we did from current SCons. Previously we have already moved the codegen part into individual python tools. This time we want to extract the build variables from SCons. Moving configuration step out of SCons doesn't increase the difficulty of moving away from SCons. Instead, it provides the opportunity to share the config between different build system. We can develop a parallel build system with the same config and then get away from SCons.

This makes a lot of sense to me. Thanks for the explanation.

> Finally, I would like to mention again that the customized build is already enabling in current SCons but seems no document talking about that. The KConfig feature is just bring that into users' eyes and making it more easily to use. We can also hide the capability but I don't think you want going that way.

"Making it more easy to use" is in the eye of the beholder :). As people who have significant experience with KConfig, we agree that it's pretty easy to use. But for people who have never seen it before, a flat file is much easier to understand.

I am starting to come around a bit more to this 😄. Especially if we have some nice documentation (like an extension of what you wrote in the response above), I think that telling users "continue to build gem5 with `build/<default build_opts>/gem5.opt` and you can customize that build by running `scons menuconfig build/<default build_opts>` make sense.

I would suggest that we try to keep the "default" and the "expert" builds as similar as possible when describing them in documentation. I.e., you can say "you can use any directory name for the build directory, but we will use `build/<variant or default>/gem5.opt` in all of our examples." Let's try not to confuse people by showing many different possible ways. I abide by the Python mantra of ["there is one way to do it"](https://peps.python.org/pep-0020/#the-zen-of-python), PEP 20 line 13.

BobbyRBruce commented:
I'd like to raise the potential issue that this may break some of our testing infrastructure. The tests run via "tests/main.py", which I call "testlib", will build the gem5 binaries they need to run the tests and I think some of the changes outlined here might break this.

The code for creating a gem5 binary for a test is here: https://github.com/gem5/gem5/blob/af72b9ba580546ac12ce05bfaac3fd53fa8699f4/tests/gem5/fixture.py#L186. As you can see it constructs the build command by utilizing the `PROTOCOL` variable and depeneds on the files "build_opts".

While most of our tests build a pretty vanilla "build/ALL/gem5.opt" binary to run tests, other target specific ISAs with specific protocols.

I think `./main.py list --length=long gem5/riscv-boot-tests` should sanity test any changes you have. Tests invoked with that command try to "build/NULL_MI_example/gem5.opt".

wmin0 commented:
Hi Bobby,

Thanks for raising that concern. I think fixing the issue is our responsibility for this PR. We will look at the issue and update the PR accordingly.

Hi Jason,

> I would suggest that we try to keep the "default" and the "expert" builds as similar as possible when describing them in documentation. I.e., you can say "you can use any directory name for the build directory, but we will use build/<variant or default>/gem5.opt in all of our examples." 

Originally our plan is not telling people there's a way to have different build directory name in current document [here](https://www.gem5.org/documentation/general_docs/building) and [here](https://www.gem5.org/documentation/learning_gem5/part1/building) to reduce the confusion. Instead we will add a whole new section under building and learning_gem5 part2 to tell about how to do customized build. We want only advanced users knowing that feature as needed.

But yes, if you think we should tell people there's a possibility. We can add your sentence to the current building document and redirect people to our new content if they want to know more detail. How do you think about that?

For the documentation, as mentioned above, we plan to support both general docs and learning gem5. We are working on a PR to website repo. Stay tune :)

Thanks.

rogerchang23424 commented:
Update tests script few minutes ago. Let it test if it can pass PR Github CI

rogerchang23424 commented:
Test: test-simpoints-se-restore-ALL-x86_64-opt Failed after test lib changes

powerjg commented:
> Test: test-simpoints-se-restore-ALL-x86_64-opt Failed after test lib changes

It's not just this change that's failing this test. I think it's an issue with the test itself. We're looking into it.

The documentation updates that you suggest sound great to me as well!

powerjg commented:
> > Test: test-simpoints-se-restore-ALL-x86_64-opt Failed after test lib changes

Maybe I was wrong about this. Here's a bit more detail on the failure I was seeing: https://github.com/gem5/gem5/pull/107#issuecomment-1645741641

I think https://github.com/gem5/gem5/pull/88 solves this problem, but it looks like that changeset is part of your branch. I'm not sure what's going on. We'll keep digging :)

rogerchang23424 commented:
Thanks @powerjg for help, now it can pass the GitHub CI.

I will start to write to document about kconfig usage, it may need to take about one and two weeks. So the PR will not update for a while. Once the document is finished, I will create another PR in gem5 website.

Feel free to ping me if you have any concern

Thanks

rogerchang23424 commented:
@gabemblack

powerjg commented:
From my side, I'm waiting for documentation on the gem5 website, then I'll provide a final review of this PR.

rogerchang23424 commented:
We are drafting the document for the kconfig website. Once completed, I will update the PR.

rogerchang23424 commented:
gem5/website#79

powerjg commented:
@rogerchang23424 thanks for the updates. I will get to reviewing this ASAP. I was out the last week or so, and I'm still catching up. Hopefully, I can review within 1 week. Ping me if you don't hear anything :).

Note: This is blocked by #252

powerjg commented on .gitignore:
```diff
@@ -32,3 +32,4 @@ configs/example/memcheck.cfg
 configs/dram/lowp_sweep.cfg
 .pyenv
 .vscode
+gem5_build
```
Just curious about why you decided to change the name of the build directory to `gem5_build` from `build`?

rogerchang23424 replied:
No, the users can still use the `build` directory to build gem5. The `gem5_build` directory is used for kconfig to avoid conflict with old scheme in `build` directory.

rogerchang23424 replied:
Removed

powerjg commented on src/learning_gem5/part3/Kconfig:
```diff
@@ -23,7 +23,10 @@
-sticky_vars.Add(BoolVariable('BUILD_GPU', 'Build the compute-GPU model',
-                             False))
+cont_choice "Ruby protocol"
+    config RubyProtocolMsi
```
Why are these `CamelCase` when all other options are all caps `SNAKE_CASE`?

What is the style rule here?

rogerchang23424 replied:
Changed to SNAKE_CASE fomat

powerjg commented on src/arch/null/Kconfig:
```diff
@@ -23,5 +23,5 @@
 
-Import('*')
-sticky_vars.Add(BoolVariable('USE_NULL_ISA', 'Enable NULL ISA support', False))
+config USE_NULL_ISA
```
Does this work as expected? `NULL` should mean that you have *no* ISAs compiled. I.e., we should be able to build gem5 without any ISAs and it should work. I don't think that `NULL` should be a kconfig option.

I'll note that I tried enabling this and got an error that `AtomicSimpleCPU` doesn't exist (or something like that).

rogerchang23424 replied:
Changed to `BUILD_ISA`

powerjg commented on src/Kconfig:
```diff
@@ -0,0 +1,52 @@
+    default ""
+
+config USE_EFENCE
+    bool "Link with Electric Fence malloc debugger"
```
What is this? Can you provide more help in the description here?

rogerchang23424 replied:
This option comes form:
https://github.com/gem5/gem5/blob/5fd901ffbb42bbb89e636e56caf5d3901e2376d3/SConsopts#L52

We just copied the config option name and description.

powerjg commented on src/Kconfig:
```diff
@@ -0,0 +1,52 @@
+    depends on BATCH
+
+config M5_BUILD_CACHE
+    string "Cache built objects in this directory"
```
What is this? Can you provide more help in the description here?

rogerchang23424 replied:
This option comes form:
https://github.com/gem5/gem5/blob/5fd901ffbb42bbb89e636e56caf5d3901e2376d3/SConsopts#L51

We just copied the config option name and description.

powerjg commented on src/Kconfig:
```diff
@@ -0,0 +1,52 @@
+
+mainmenu "$(MAIN_MENU_TEXT)"
+
+config BATCH
+    bool "Use batch pool for build and test"
+    default n
+
+config BATCH_CMD
+    string "Batch pool submission command name"
+    default "qdo"
+    depends on BATCH
```
Maybe in another changeset we should remove this. I believe this has bit-rotted beyond usefulness.

rogerchang23424 replied:
BATCH
https://github.com/gem5/gem5/blob/5fd901ffbb42bbb89e636e56caf5d3901e2376d3/SConsopts#L49

BATCH_CMD
https://github.com/gem5/gem5/blob/5fd901ffbb42bbb89e636e56caf5d3901e2376d3/SConsopts#L50

> Maybe in another changeset we should remove this. I believe this has bit-rotted beyond usefulness.

They are existing in the origin `SConsopts` file. We should open another PR to discuss about the options usage.

powerjg commented on SConstruct:
```diff
@@ -205,7 +196,67 @@ if not ('CC' in main and 'CXX' in main):
+Targets:
+        To build gem5 using a predefined configuration, use a target with
+        a directory called "build" in the path, followed by a directory named
+        after a predefined configuration, and then the actual target, likely
```
You should specify that these are found in `build_opts`

rogerchang23424 replied:
included

powerjg commented on SConstruct:
```diff
@@ -205,7 +196,67 @@ if not ('CC' in main and 'CXX' in main):
+        after a predefined configuration, and then the actual target, likely
+        a gem5 binary. For example:
+
+        scons build/X86/gem5.opt
```
```suggestion
        scons build/ALL/gem5.opt
```
Let's default to `ALL`

powerjg commented on SConstruct:
```diff
@@ -205,7 +196,67 @@ if not ('CC' in main and 'CXX' in main):
+            disabled.
+
+        gem5 can also be built as a static or dynamic library. In that case,
+        the extension is fixed by the operating system, so the binary type
```
> the extension is fixed by the operating system

What does this mean? Do we support any OSes that use an extension other than `.so`?

rogerchang23424 replied:
Yes, the users should determine the extension name beforehand to build gem5 dynamic library. For example, `.so` for linux platform and `.dylib` for MacOS. I have updated the help description to include the build commands for Linux and MacOS respectively. The users should run the corresponding build command in their OS.

powerjg commented on SConstruct:
```diff
@@ -215,52 +266,138 @@ Default(environ.get('M5_DEFAULT_BINARY', 'build/ARM/gem5.debug'))
+
+        The tools available for working with kconfig are generally very
+        similar to ones used with the linux kernel, so information about the
+        kernel versions will typically (but not always) apply here as well.
```
> will typically (but not always) apply here

When does kernel documentation *not* apply here? That's going to be important to document.

rogerchang23424 replied:
Deleted

powerjg commented on SConstruct:
```diff
@@ -215,52 +266,138 @@ Default(environ.get('M5_DEFAULT_BINARY', 'build/ARM/gem5.debug'))
+        to configure, and then the name of the tool. For example, to run
+        menuconfig on directory build/foo/bar, run:
+
+        scons menuconfig build/foo/bar
```
> `build/foo/bar`

All of these examples make it seem like you have to use `build` as the root of the directory structure. I don't think this is the case, correct?

rogerchang23424 replied:
After the kconfig changed, the user can build gem5 without need to follow the naming structure of `build/<build_opts>/gem5.opt`. Instead, they can use the customized the build namiing like `gem5_build/gem5.opt`, `foo_bar/gem5.debug` etc.

powerjg commented on SConstruct:
```diff
@@ -680,59 +821,13 @@ for variant_path in variant_paths:
+    extras_file = os.path.join(gem5_build, 'extras')
+    extras_var = Variables(extras_file, args=ARGUMENTS)
+
+    extras_var.Add(('EXTRAS', 'Add extra directories to the compilation', ''))
```
`EXTRAS` is missing from all of the new documentation. Is that on purpose?

I think we should at least mention how `EXTRAS` can be used now.

rogerchang23424 replied:
Please see the commit message of this 
https://github.com/gem5/gem5/pull/69/commits/0ee2917562e3e234cda5b1921ae9c5d9656bd6f6

```
Switch SCons/gem5 to use Kconfig for configuration, except EXTRAS which
is still a sticky SCons variable. This is necessary because EXTRAS also
controls what config options exist. If it came from Kconfig itself, then
there would be a circular dependency. This dependency could
theoretically be handled by reparsing the Kconfig when EXTRAS
directories were added or removed, but that would be complicated, and
isn't supported by kconfiglib. It wouldn't be worth the significant
effort it would take to add it, just to use Kconfig more purely.
```

powerjg commented on SConstruct:
```diff
@@ -770,24 +876,50 @@ for variant_path in variant_paths:
+            kconfig.setconfig(env, kconfig_file.abspath, config_file.abspath,
+                    ARGUMENTS)
+        else:
+            error(f'Unrecognized kconfig action {kconfig_action}')
```
I don't think it's possible to hit this line of code because `kconfig_action` will be false if it's not one of the above. I guess there's nothing to change here, but it's a bit strange.

powerjg reviewed (commented):
Another random thing I just realized... now building the "m5 tools" will have a different method than building gem5 since it will not use kconfig. I guess this is OK since it already had some quirks that made it different than building gem5.

powerjg commented:
BTW, I don't know how/why it was merged, but the file `KCONFIG.md` should probably be updated. See https://gem5-review.googlesource.com/c/public/gem5/+/58634

rogerchang23424 committed:
scons: Update the Kconfig build options

The CL updates the Kconfig:
1. Replace the USE_NULL_ISA with BUILD_ISA
2. The USE_XXX_ISAs are depends on BUILD_ISA
3. If the BUILD_ISA is set, at least one of USE_XXX_ISAs must be set
4. Refactor the USE_KVM option

Change-Id: I2a600dea9fb671263b0191c46c5790ebbe91a7b8

rogerchang23424 committed:
scons: Update Kconfig description

Change-Id: I69206fb9881bc0d53660bbd1cf8fc225ead9fea3

powerjg commented:
From my point of view, there's two small things on the to do list before we merge this.
1. I've asked at least one person unfamiliar with the patch to try it out and see if they understand things based on the documentation added to the website.
2. We need to have a changeset that allows all of the tests to pass.

rogerchang23424 committed:
scons: Refactor the USE_SYSTEMC option

Change-Id: I2f51081e0db932b83eea9dd395551afe13d54a34

rogerchang23424 committed:
util: Add python3-tk package to dockerfile

The guiconfig required the python3-tk package to run.

Change-Id: I1d126021c2c57448b1ceefb9fff256e2a6bbbf33

rogerchang23424 committed:
tests: Update Gem5Fixture to compatible of Kconfig system

Used after the build system changed to Kconfig

Change-Id: I699b36f09691dc821da8ee80fe5b60f30fe2179c

rogerchang23424 committed:
misc: Update daily-test.yaml to match Kconfig build system configuration

Change-Id: I20b04d006c67374e3226a91c03550f2731ed7fe7

rogerchang23424 commented:
Since the PR changes the methods to specify config options, I have created the other PR https://github.com/gem5/gem5/pull/478 to ensure the daily tests can work well after kconfig changed.

rogerchang23424 committed:
scons: Add new config option HAVE_CAPSTONE to Kconfig

The config option HAVE_CAPSTONE is added in the previous [1] and
the Kconfig options should be sync with it.

[1] https://github.com/gem5/gem5/pull/494

Change-Id: Id83718bc825f53d87d37d6ac930b96371209bdb3

rogerchang23424 commented:
Hi maintainers,



> I'd like to raise the potential issue that this may break some of our testing infrastructure. The tests run via "tests/main.py", which I call "testlib", will build the gem5 binaries they need to run the tests and I think some of the changes outlined here might break this.
> 
> The code for creating a gem5 binary for a test is here:
> 
> https://github.com/gem5/gem5/blob/af72b9ba580546ac12ce05bfaac3fd53fa8699f4/tests/gem5/fixture.py#L186
> 
> . As you can see it constructs the build command by utilizing the `PROTOCOL` variable and depeneds on the files "build_opts".
> While most of our tests build a pretty vanilla "build/ALL/gem5.opt" binary to run tests, other target specific ISAs with specific protocols.
> 
> I think `./main.py list --length=long gem5/riscv-boot-tests` should sanity test any changes you have. Tests invoked with that command try to "build/NULL_MI_example/gem5.opt".

Hi maintainers,

To fix the issue of daily tests after the kconfig build system introduced, I have added some commits to fix daily tests and test in #478. The test works well as the action(https://github.com/gem5/gem5/actions/runs/6602383103) with the current develop branch. Cloud we start **merge process** of the PR?

Thanks

powerjg commented:
> Could we start the merge process of the PR?

There's one more potential blocker and one other small bump:
1. What about the [weekly tests](https://github.com/gem5/gem5/blob/stable/.github/workflows/weekly-tests.yaml) and the [GPU tests](https://github.com/gem5/gem5/blob/stable/.github/workflows/gpu-tests.yaml)? I *think* the tests are fine, but I don't fully understand the new build system. 
2. Bobby is on vacation this week. Unfortunately, he is the only person who knows how to merge changes to the GitHub Actions workflows. There is a complication as these files have to be on the *stable* branch instead of develop. 

Could you look into (1) above and then ping Bobby in 1 week to merge?

rogerchang23424 commented:
> What about the [weekly tests](https://github.com/gem5/gem5/blob/stable/.github/workflows/weekly-tests.yaml) and the [GPU tests](https://github.com/gem5/gem5/blob/stable/.github/workflows/gpu-tests.yaml)? I think the tests are fine, but I don't fully understand the new build system.

PR for weekly tests and gpu tests: https://github.com/gem5/gem5/pull/503

> Could you look into (1) above and then ping Bobby in 1 week to merge?

Sure

powerjg commented:
@rogerchang23424, I want to talk about this during the gem5 developers meeting tomorrow. Could you let me know what the current state is and if there's any blockers?

BTW, I apologize for the bad time zone for you! We will try to schedule this at a better time in the future.

rogerchang23424 commented:
> I want to talk about this during the gem5 developers meeting tomorrow. Could you let me know what the current state is and if there's any blockers?

It OK for meet to join the this meeting tomorrow.

> BTW, I apologize for the bad time zone for you! We will try to schedule this at a better time in the future.

The meeting time is 1:00 am Nov. 10 in Taiwan time zone. I think schedule the meeting in the afternoon(Pacific standard time) will be better for me. It's morning in Taiwan time.

BobbyRBruce commented on site_scons/gem5_scons/kconfig.py:
```diff
@@ -0,0 +1,183 @@
+}
+
+
+def _prep_env(env, base_kconfig, config_path=None):
```
This goes for all the functions in this file. I think this is raelly important to document and get right, so could i trouble you to add typing information here and a docstring explaining the the function. A typical example in gem5 would be:

```py
def get_isa_from_str(input: str) -> ISA:
    """
    Will return the correct enum given the input string. This is matched on
    the enum's value. E.g., "x86" will return ISA.X86. Throws an exception if
    the input string is invalid.

    `get_isas_str_set()` can be used to determine the valid strings.

    This is for parsing text inputs that specify ISA targets.

    :param input: The ISA to return, as a string. Case-insensitive.
    """
```

Not the type for each parameter, the return type, and a Sphinx doctstring explaining the function. Yours needn't be as detailed. A sentence will do.

rogerchang23424 replied:
Added in the following function

BobbyRBruce commented on src/arch/arm/fastmodel/SConsopts:
```diff
@@ -1,5 +1,4 @@
-# Copyright 2019 Google, Inc.
-# All rights reserved.
+# Copyright 2022 Google LLC
```
I notice you keep using 2022 in some of these files. If this is intentional, that's fine, but it's 2023.

I also imagine the 2019 copyright date may still be valid?: `Copyright 2019-2023 Google LLC`?

BobbyRBruce commented on src/systemc/Kconfig:
```diff
@@ -0,0 +1,33 @@
+    depends on HAVE_SYSTEMC
+    bool "Enable SystemC API support"
+    default y if HAVE_SYSTEMC
+    default n
```
Why do we have two defaults here? The default is `y` if `HAVE_SYSTEMC` is true, but if `HAVE_SYSTEMC` is false this whole y/n option is not available. So why have a `default n`? It's already not there in this case.

BobbyRBruce commented on util/dockerfiles/ubuntu-22.04_all-dependencies/Dockerfile:
```diff
@@ -31,6 +31,6 @@ RUN apt -y update && apt -y upgrade && \
     libprotobuf-dev protobuf-compiler libprotoc-dev libgoogle-perftools-dev \
     python3-dev doxygen libboost-all-dev libhdf5-serial-dev python3-pydot \
-    libpng-dev libelf-dev pkg-config pip python3-venv black
+    libpng-dev libelf-dev pkg-config pip python3-venv black python3-tk
```
If you don't already have a PR do do so, update the website build dependencies to highlight `python3-tk` is needed as it contains kconfiglib

rogerchang23424 replied:
The `python3-tk` is only required for the `guiconfig.py` of `Kconfiglib`. Do you think is better to create another PR for the updating the `Dockerfile` for `python3-tk`?

BobbyRBruce replied:
One PR is my preference. I like these KConfig commits all in one place.

rogerchang23424 replied:
> If you don't already have a PR do do so, update the website build dependencies to highlight python3-tk is needed as it contains kconfiglib

Update the document in gem5/website#86

BobbyRBruce commented on tests/gem5/suite.py:
```diff
@@ -45,7 +45,7 @@
 from testlib.helper import log_call
 from testlib.configuration import constants, config
-from .fixture import TempdirFixture, Gem5Fixture, VariableFixture
+from .fixture import TempdirFixture, Gem5KconfigFixture, VariableFixture
```
Do we still need `Gem5Fixture` for anything? Doesn't `Gem5KconfigFixure` supercede it? If we don't then please get rid `Gem5Fixture`?.

rogerchang23424 replied:
Changed back to the Gem5Fixture

BobbyRBruce reviewed (changes requested):
All very minor, mostly questions. This is great work. Sorry for taking so long to get round to it.

rogerchang23424 committed:
util: Update the tlm README

Change-Id: I4006257bf55d7065136347788783796fd39ab725

rogerchang23424 committed:
util: Update the gem5_within_systemc README

Change-Id: Ife34fe5ccd00fa2c6a83f34af49333d49017dfed

rogerchang23424 committed:
fastmodel: Simply the logic of USE_ARM_FASTMODEL setting

Change-Id: Ib00cf83ca881727987050a987a2adb1e9f9d31ef

BobbyRBruce reviewed (dismissed):
Ok. Looks fine to me. Let's get this in.

BobbyRBruce commented:
I'll give it until tomorrow to merge, but just to confirm @rogerchang23424 : Are you ok with everything?

rogerchang23424 commented:
> I'll give it until tomorrow to merge, but just to confirm @rogerchang23424 : Are you ok with everything?

Sure

BobbyRBruce committed:
Merge branch 'develop' into develop-kconfig

rogerchang23424 commented:
Thanks for merging the PR

EarlOu commented:
Thanks Jason, Bobby, Roger and also Gabe! It's a long journey to get this in! Also thanks everyone who contributes to the discussion!
